src/Features/Core/Portable/PullMemberUp/AbstractMembersPullerService.cs

```diff
@@ -9,23 +9,33 @@
 using System.Linq;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.CodeAnalysis.AddImports;
 using Microsoft.CodeAnalysis.CodeActions;
 using Microsoft.CodeAnalysis.CodeGeneration;
 using Microsoft.CodeAnalysis.Editing;
+using Microsoft.CodeAnalysis.LanguageServices;
 using Microsoft.CodeAnalysis.PullMemberUp;
+using Microsoft.CodeAnalysis.RemoveUnnecessaryImports;
 using Microsoft.CodeAnalysis.Shared.Extensions;
 using Microsoft.CodeAnalysis.Shared.Utilities;
 using Roslyn.Utilities;
 using static Microsoft.CodeAnalysis.CodeActions.CodeAction;
 
 namespace Microsoft.CodeAnalysis.CodeRefactorings.PullMemberUp
 {
-    internal static class MembersPuller
+    internal abstract class AbstractMembersPullerService<TUsingOrAliasSyntax> : IMembersPullerService
+        where TUsingOrAliasSyntax : SyntaxNode
     {
         /// <summary>
-        /// Return the CodeAction to pull <paramref name="selectedMember"/> up to destinationType. If the pulling will cause error, it will return null.
+        /// Gets all the imports/aliases in scope of <paramref name="node"/>.
         /// </summary>
-        public static CodeAction TryComputeCodeAction(
+        protected abstract ImmutableArray<TUsingOrAliasSyntax> GetImports(SyntaxNode node);
+
+        protected const string Kind = "PullMemberRemovableImport";
+
+        protected static SyntaxAnnotation Annotation = new(Kind);
```

Comment:
why protected?

Reply:
I remember thinking that I didn't want this to be confused for some type of annotation that was used to mark members, and I was thinking it would be exclusively for use within the class. Maybe I made it protected so the individual languages could access it, although that shouldn't be necessary now that the annotations are added and accessed in the abstract class. Do you think I should maybe give the variables a different, more descriptive name and make them internal, or should I just make it private?

Reply:
Almost always make your code as restrictive as necessary :)

IN terms of naming, feel free to rename if you think of something better.  However, as there is only one use case of this, s_annotation would be fine (With a descriptive doc comment explaining what it is for).